smithy-smoke-test-traits/src/main/java/software/amazon/smithy/smoketests/traits/SmokeTestCaseValidator.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.smoketests.traits;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.OperationIndex;
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.model.node.ObjectNode;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.NodeValidationVisitor;
+import software.amazon.smithy.model.validation.ValidationEvent;
+import software.amazon.smithy.model.validation.node.TimestampValidationStrategy;
+
+/**
+ * Validates the following:
+ *
+ * <ul>
+ *     <li>vendorParamsShape is a valid shape.</li>
+ *     <li>vendorParams is compatible with vendorParamsShape.</li>
+ *     <li>input is valid for the operation under test.</li>
+ * </ul>
+ */
+public class SmokeTestCaseValidator extends AbstractValidator {
+
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        OperationIndex operationIndex = OperationIndex.of(model);
+        List<ValidationEvent> events = new ArrayList<>();
+        for (Shape shape : model.getShapesWithTrait(SmokeTestsTrait.class)) {
+            SmokeTestsTrait trait = shape.expectTrait(SmokeTestsTrait.class);
+            List<SmokeTestCase> testCases = trait.getTestCases();
+
+            for (int i = 0; i < testCases.size(); i++) {
+                SmokeTestCase testCase = testCases.get(i);
+
+                // Validate vendor params shape
+                Optional<ShapeId> vendorParamsShapeOptional = testCase.getVendorParamsShape();
+                Optional<ObjectNode> vendorParamsOptional = testCase.getVendorParams();
+                if (vendorParamsShapeOptional.isPresent()) {
+                    if (!vendorParamsOptional.isPresent()) {
+                        events.add(warning(shape, trait,
+                                "Smoke test case defined a `vendorParamsShape` but no `vendorParams`"));
```

Comment:
We should put a warning for the case where they have vendor params but no shape too

Reply:
Done